client/consensus/grandpa/authorities.go

```diff
@@ -0,0 +1,611 @@
+	return pc.canonHeight + pc.delay
+}
+
+// AuthoritySetChanges Tracks historical authority set changes. We store the block numbers for the last block
+// of each authority set, once they have been finalized. These blocks are guaranteed to
+// have a justification unless they were triggered by a forced hashNumber.
+type AuthoritySetChanges[N constraints.Unsigned] []authorityChange[N]
+
+// append an authorityChange to AuthoritySetChanges
+func (asc *AuthoritySetChanges[N]) append(setId uint64, blockNumber N) {
+	*asc = append(*asc, authorityChange[N]{
+		setId:       setId,
+		blockNumber: blockNumber,
+	})
+}
+
+// Three states that can be returned: Latest, Set (tuple), Unknown
+// Latest => bool
+// Set => &AuthorityChange
+// Unknown => nil
+// TODO for reviewers, this can be a VDT but I'm not sure its needed
+func (asc *AuthoritySetChanges[N]) getSetID(blockNumber N) (latest bool, set *authorityChange[N], err error) {
+	if asc == nil {
+		return false, nil, fmt.Errorf("getSetID: authSetChanges is nil")
+	}
+	authSet := *asc
+	last := authSet[len(authSet)-1]
+	if last.blockNumber < blockNumber {
+		return true, nil, nil // Latest case
+	}
+
+	idx, _ := searchSetChanges(blockNumber, authSet)
+	if idx < len(authSet) {
+		authChange := authSet[idx]
+
+		// if this is the first index but not the first set id then we are missing data.
+		if idx == 0 && authChange.setId != 0 {
+			return false, nil, nil // Unknown case
+		}
+
+		return false, &authChange, nil // Set case
+	}
+
+	return false, nil, nil // Unknown case
+}
+
+func (asc *AuthoritySetChanges[N]) insert(blockNumber N) {
+	var idx int
+	if asc == nil {
+		panic("authority set changes must be initialized")
+	} else {
+		idx, _ = searchSetChanges(blockNumber, *asc)
+	}
+
+	set := *asc
+
+	var setId uint64
+	if idx == 0 {
+		setId = 0
+	} else {
+		setId = set[idx-1].setId + 1
+	}
+
+	if idx != len(set) && set[idx].setId == setId {
+		panic("inserting authority set hashNumber")
+	}
+
+	change := authorityChange[N]{
+		setId:       setId,
+		blockNumber: blockNumber,
+	}
+
+	// Insert hashNumber at index
+	if len(set) <= idx {
+		set = append(set, change)
+	} else {
+		set = append(set[:idx+1], set[idx:]...)
+		set[idx] = change
+	}
+	*asc = set
+}
+
+// This logic is used in warp sync proof
+func (asc *AuthoritySetChanges[N]) iterFrom(blockNumber N) *AuthoritySetChanges[N] {
+	if asc == nil {
+		return nil
+	}
+	authSet := *asc
+
+	idx, found := searchSetChanges(blockNumber, *asc)
+	if found {
+		// if there was a hashNumber at the given block number then we should start on the next
+		// index since we want to exclude the current block number
+		idx += 1
+	}
+
+	if idx < len(*asc) {
+		authChange := authSet[idx]
+
+		// if this is the first index but not the first set id then we are missing data.
+		if idx == 0 && authChange.setId != 0 {
+			return nil
+		}
+	}
+
+	iterChanges := authSet[idx:]
+	return &iterChanges
+}
+
+////// Additional types //////
```

Comment:
I would suggest to move those types definitions to the top of the file
```suggestion
```

Reply:
Yes that makes sense to me. I had initially not done this to try and maintain substrates file structure as close as I could, but I think its cleaner to have on top